### DISCUSSION TITLE
关于IntegralCast的问题

### DISCUSSION THREAD

**Discussion by ghost** [arcsysu/SYsU-lang#29] [Q&A] [answered]
在`tester/h_functional/100_int_literal.sysu.c`中，类似这样的超过int范围的变量赋值会发生隐式类型转换
```
const int k0 = -2147483648;
const int k1 = 0x80000000;
const int k2 = 0x80000000 + 1;
const int k3 = 0x7fFffffF;
const int k4 = 0x7fFffffF - 1;
```
其中`const int k2`的子树部分如下
```
|-DeclStmt 0xacd128 <line:111:3, col:32>
| `-VarDecl 0xacd030 <col:3, col:31> col:13 k2 'const int' cinit
|   `-ImplicitCastExpr 0xacd110 <col:18, col:31> 'int' <IntegralCast> <-----------------------------------------------
|     `-BinaryOperator 0xacd0f0 <col:18, col:31> 'unsigned int' '+'
|       |-IntegerLiteral 0xacd098 <col:18> 'unsigned int' 2147483648
|       `-ImplicitCastExpr 0xacd0d8 <col:31> 'unsigned int' <IntegralCast> <-----------------------------------------------
|         `-IntegerLiteral 0xacd0b8 <col:31> 'int' 1
```
好像这里ImplicitCastExpr涉及了很多语义信息，需要判断左值、右值的类型，而非简单嵌套ImplicitCastExpr即可。不知道该如何做到和clang一样的输出，请问助教是否遇到这样的问题，有没有什么好方法解决呢？
或者可否把这几个改成int范围内的数据，简单一些呢：）

**Comment by Yorkking** [comment, marked as answer]
`100_int_literal.sysu.c` 的问题，这里涉及到类型转换的问题，unsigned int 和 int 转换的问题。观察了一下 clang 的语法树得出结论：

- 1. a = b + c： b 和 c 类型不一致时，往类型大的转，比如 int ->unsigned int
- 2. a 和 b + c 返回的结果类型不一致时，强行往 a 的类型转.。
个人的解决方法是，给每个节点引入一个额外的 type 属性，比如问题中的 `const int k2 = 0x80000000 + 1;` ,  首先 `0x80000000` 加入 `"type:unsigned int"` 属性, `1` 加入 `"type:int"`  属性。然后在二元表达式运算的时候，比较两个的 `type` 是否相同，然后按照上述的规则进行类型转换，即插入 `ImplicitCastExpr`。然后变量声明的时候也做类似的检查 `type` 操作。

另外，实际上只有这个例子会涉及到复杂的类型转换，只需要针对该例子修改和 `const int k2 = 0x80000000 + 1;` 相关的文法即可。
